File: webapp/packages/lyrics-source/src/convert.ts
export function lrcToVtt(lines: Array<{ time?: number; text: string }>) {
  // lines: time in seconds
  const out: string[] = []
  out.push('WEBVTT')
  out.push('')
  let idx = 0
  for (const line of lines) {
    const start = typeof line.time === 'number' ? secondsToStamp(line.time) : '00:00.000'
    const end = typeof line.time === 'number' ? secondsToStamp((line.time as number) + 3) : '00:03.000'
    out.push(`${idx}`)
    out.push(`${start} --> ${end}`)
    out.push(line.text)
    out.push('')
    idx++
  }
  return out.join('\n')
}

function secondsToStamp(s: number) {
  const hh = Math.floor(s / 3600)
  const mm = Math.floor((s % 3600) / 60)
  const ss = Math.floor(s % 60)
  const ms = Math.floor((s - Math.floor(s)) * 1000)
  return `${pad(hh)}:${pad(mm)}:${pad(ss)}.${String(ms).padStart(3, '0')}`
}

function pad(n: number) { return String(n).padStart(2, '0') }

export function addVttToJcrd(jcrd: any, vttPath: string) {
  jcrd = jcrd || {}
  jcrd.assets = jcrd.assets || {}
  jcrd.assets.lyrics_vtt = vttPath
  return jcrd
}

File: services/lyrics_source/convert.py
from typing import List, Dict, Any

def seconds_to_stamp(s: float) -> str:
    hh = int(s // 3600)
    mm = int((s % 3600) // 60)
    ss = int(s % 60)
    ms = int((s - int(s)) * 1000)
    return f"{hh:02d}:{mm:02d}:{ss:02d}.{ms:03d}"

def lrc_to_vtt(lines: List[Dict[str, Any]]) -> str:
    out = ["WEBVTT", ""]
    idx = 0
    for ln in lines:
        start = ln.get('time')
        if start is None:
            start_s = 0.0
        else:
            start_s = float(start)
        end_s = start_s + 3.0
        out.append(str(idx))
        out.append(f"{seconds_to_stamp(start_s)} --> {seconds_to_stamp(end_s)}")
        out.append(ln.get('text', ''))
        out.append("")
        idx += 1
    return "\n".join(out)

def add_vtt_to_jcrd(jcrd: Dict[str, Any], vtt_path: str) -> Dict[str, Any]:
    j = dict(jcrd or {})
    j.setdefault('assets', {})
    j['assets']['lyrics_vtt'] = vtt_path
    return j
